components/vocabularies/oncotree/api/src/main/java/org/phenotips/vocabulary/internal/solr/OncoTree.java

```diff
@@ -128,34 +146,97 @@ InputStream getInputStream(@Nonnull final URL url) throws IOException
     private void processDataRow(@Nonnull final CSVRecord row, @Nonnull final Collection<SolrInputDocument> data)
     {
         final SolrInputDocument doc = new SolrInputDocument();
-
         for (int i = 0; i < row.size(); i++) {
-            final String label = COLUMNS[i];
             final String value = row.get(i);
-            applyWriteStrategies(doc, label, value);
+            if (StringUtils.isNotBlank(value)) {
+                if (i == 0) {
+                    processTissue(doc, COLUMNS[i], value);
+                } else if (i < 5) {
+                    processName(doc, value);
+                } else {
+                    doc.addField(COLUMNS[i], value);
+                }
+            }
         }
         data.add(doc);
     }
 
     /**
-     * Writes the {@code label} and {@code value} data into the {@code doc}, as per some predefined write strategies.
+     * Processes the {@code value cancer name value}, and writes the extracted identifier and name information into
+     * {@code doc}.
      *
-     * @param doc the {@link SolrInputDocument} where the data will be written
-     * @param label the property label
-     * @param value the property value
+     * @param doc the {@link SolrInputDocument} into which data is written
+     * @param value the provided raw cancer name
      */
-    private void applyWriteStrategies(
-        @Nonnull final SolrInputDocument doc,
-        @Nonnull final String label,
-        @Nullable final String value)
+    private void processName(@Nonnull final SolrInputDocument doc, @Nonnull final String value)
+    {
+        updateParents(doc);
+        updateCancerId(doc, value);
+        updateCancerName(doc, value);
+    }
+
+    /**
+     * Tries to extract the name of the cancer from the provided raw {@code value name} string, and writes it into
+     * {@code doc}.
+     *
+     * @param doc the {@link SolrInputDocument} into which data is written
+     * @param value the provided raw cancer name
+     */
+    private void updateCancerName(@Nonnull final SolrInputDocument doc, @Nonnull final String value)
     {
-        final Collection<OncoTreePropertyWriteStrategy> strategies = STRATEGIES.get(label);
-        for (final OncoTreePropertyWriteStrategy strategy : strategies) {
-            strategy.execute(doc, label, value);
+        final String name = StringUtils.substringBefore(value, "(").trim();
```

Comment:
StringUtils.substringBefore will return the string as is if it fails (interestingly enough, that's not the case with StringUtils.substringBetween).